In [2]:
import pandas as pd
import wget
import numpy as np

## Q1

In [3]:
print(pd.__version__)

3.0.6


## Q2

In [9]:
car_fuel = pd.read_csv('car_fuel_efficiency_2026.csv')
print(car_fuel.head())
print(car_fuel.count())

   model_year  origin fuel_type         drivetrain  num_doors  \
0        2006  Europe  Gasoline  Front-wheel drive          4   
1        2008  Europe    Diesel  Front-wheel drive          4   
2        1996    Asia  Gasoline  Front-wheel drive          5   
3        1989  Europe  Gasoline  Front-wheel drive          4   
4        1994     USA    Diesel  Front-wheel drive          3   

   engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                 2180              6       243.0            3870   
1                 2390              6       272.0            4210   
2                 2320              6       267.0            4240   
3                 2130              6       258.0            4490   
4                 2580              7       304.0            4510   

   acceleration  fuel_efficiency_mpg  
0           NaN                 31.9  
1           NaN                 31.3  
2          17.3                 27.5  
3          18.7                 28.5  

There are 10000 records (rows)*

## Q3

In [11]:
print(car_fuel['fuel_type'].unique())

<StringArray>
['Gasoline', 'Diesel', 'Hybrid']
Length: 3, dtype: str


There are 3 types of fuel (Gasoline, Diesel and Hybrid)*

## Q4

In [13]:
print(car_fuel.isnull().sum())

model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors                0
engine_displacement      0
num_cylinders            0
horsepower             877
vehicle_weight           0
acceleration           264
fuel_efficiency_mpg      0
dtype: int64


There are 2 columns with missing values (horsepower and acceleration)*

## Q5

In [14]:
asia_cars = car_fuel[car_fuel['origin'] == 'Asia']
print(asia_cars['fuel_efficiency_mpg'].max())

41.2


The max fuel efficiency in asia is 41.2*

## Q6

In [16]:
print(car_fuel['horsepower'].median())
print(car_fuel['horsepower'].mode())

254.0
0    252.0
Name: horsepower, dtype: float64


In [18]:
print(car_fuel['horsepower'].fillna(252.0).median())

252.0


The median horsepower decreased (from 254.0 to 252.0)*

## Q7

In [39]:
asian_weight_year = asia_cars[['vehicle_weight','model_year']]
first_7 = asian_weight_year.head(7)
X = first_7.values
XTX = X.T.dot(X)
inv_XTX = np.linalg.inv(XTX)
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
w = inv_XTX.dot(X.T).dot(y)
print(w.sum())

0.36919696904925486


The sum of all elements of the result is approximately 0.369*